script/testing/util/test_server.py

```diff
@@ -1,162 +1,207 @@
 
-        # run the pre test tasks
         test_case.run_pre_test()
 
-        # start a thread to collect the memory info if needed
+        collect_mem_thread = None
         if self.collect_mem_info:
-            # spawn a thread to collect memory info
-            self.collect_mem_thread = PeriodicTask(
+            # Initialize memory information dict.
+            update_mem_info(self.db_instance.db_process.pid,
+                            self.incremental_metric_freq,
+                            test_case.mem_metrics.mem_info_dict)
+            # Start a thread to collect memory information periodically.
+            collect_mem_thread = PeriodicTask(
                 self.incremental_metric_freq, update_mem_info,
                 self.db_instance.db_process.pid, self.incremental_metric_freq,
                 test_case.mem_metrics.mem_info_dict)
-            # collect the initial memory info
-            update_mem_info(self.db_instance.db_process.pid, self.incremental_metric_freq,
-                            test_case.mem_metrics.mem_info_dict)
+            collect_mem_thread.start()
 
-        # run the actual test
-        with open(test_case.test_output_file, "a+") as test_output_fd:
-            ret_val, _, _ = run_command(test_case.test_command,
-                                        test_case.test_error_msg,
-                                        stdout=test_output_fd,
-                                        stderr=test_output_fd,
-                                        cwd=test_case.test_command_cwd)
-
-        # stop the thread to collect the memory info if started
-        if self.collect_mem_info:
-            self.collect_mem_thread.stop()
+        ret_val = constants.ErrorCode.ERROR
+        try:
+            with open(test_case.test_output_file, "a+") as test_output_fd:
+                ret_val, _, _ = run_command(test_case.test_command,
+                                            stdout=test_output_fd,
+                                            stderr=test_output_fd,
+                                            cwd=test_case.test_command_cwd)
+        finally:
+            if self.collect_mem_info and collect_mem_thread is not None:
+                collect_mem_thread.stop()
 
-        # run the post test tasks
-        test_case.run_post_test()
-        self.db_instance.delete_wal()
+            test_case.run_post_test()
+            self.db_instance.delete_wal()
 
         return ret_val
 
     def run(self, test_suite):
-        """ Orchestrate the overall test execution """
+        """
+        Orchestrate the overall execution of the specified test suite.
+
+        Parameters
+        ----------
+        test_suite : [TestCase]
+            A list of test cases to be run.
+
+        Returns
+        -------
+        suite_result : constants.ErrorCode
+            SUCCESS if all the tests succeeded. ERROR otherwise.
+        """
+        # TODO(WAN): I dislike this magic...
```

Comment:
Agreed. It should be easy enough to find the instances where we call the function with a single suite and wrap them in brackets.

Reply:
done